src/engine/input/key_mapping/InputEventItem.cs

```diff
@@ -0,0 +1,210 @@
+
+        if ((@event is InputEventMouseButton inputMouse) && inputMouse.ButtonIndex == (int)ButtonList.Left &&
+            xbutton.IsHovered())
+        {
+            Delete();
+            return;
+        }
+
+        if (wasPressingButton)
+        {
+            wasPressingButton = false;
+            return;
+        }
+
+        if (@event is InputEventKey iek)
+        {
+            switch (iek.Scancode)
+            {
+                case (uint)KeyList.Escape:
+                    if (AssociatedEvent == null)
+                    {
+                        Delete();
+                        return;
+                    }
+
+                    InputGroupList.Instance.WasListeningForInput = true;
+                    WaitingForInput = false;
+                    UpdateButtonText();
+                    return;
+                case (uint)KeyList.Alt:
+                case (uint)KeyList.Shift:
+                case (uint)KeyList.Control:
+                    return;
+            }
+        }
+
+        // The old godot input event. Null if this event is assigned a value the first time.
+        var old = AssociatedEvent;
+        AssociatedEvent = new ThriveInputEventWithModifiers((InputEventWithModifiers)@event);
+
+        // Get the conflicts with the new input.
+        var conflict = InputGroupList.Instance.Conflicts(this);
+        if (conflict != null)
+        {
+            AssociatedEvent = old;
+
+            // If there are conflicts detected reset the changes and ask the user.
+            InputGroupList.Instance.ShowInputConflictDialog(this, conflict, (InputEventWithModifiers)@event);
+            return;
+        }
+
+        // Check if the input is already defined for this action
+        // This code works by finding a pair
+        for (var i = 0; i < AssociatedAction.Inputs.Count; i++)
+        {
+            for (var x = i + 1; x < AssociatedAction.Inputs.Count; x++)
+            {
+                if (AssociatedAction.Inputs[i].Equals(AssociatedAction.Inputs[x]))
+                {
+                    // Pair found (input already defined)
+                    Delete();
+                    return;
+                }
+            }
+        }
+
+        WaitingForInput = false;
+        JustAdded = false;
+        InputGroupList.Instance.WasListeningForInput = false;
+
+        // Update the godot InputMap
+        InputGroupList.Instance.ControlsChanged();
+
+        // Update the button text
+        UpdateButtonText();
+    }
+
+    public override bool Equals(object obj)
+    {
+        if (!(obj is InputEventItem input))
+            return false;
+
+        return Equals(AssociatedEvent, input.AssociatedEvent);
+    }
+
+    public override int GetHashCode()
+    {
+        return AssociatedEvent.GetHashCode();
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        AssociatedAction = null;
+        button?.Dispose();
+        xbutton?.Dispose();
+        base.Dispose(disposing);
```

Comment:
This doesn't seem to dispose any classes that want to be disposed, so can't this be removed?

Reply:
the important thing in the dispose method is the 
`AssociatedAction = null;`
line.
Everything else is to satisfy CI

Reply:
Setting AssociatedAction to null is important for the GC because otherwise I would have an unhandled reference loop

Reply:
As long as you don't reference it externally, it should get deleted? Isn't C# smart enough that it handles reference loops that aren't referenced from outside? Or you could put in a weak reference into the AssociatedAction or make AssociatedAction a weak reference.

Reply:
C#'s GC odd being smart enough, but I've experienced that loops can still cause memory problems. Just fixing this in the Dispose is the easiest way to make sure there aren't any problems

Reply:
Anyone doing further changes needs to be careful with this class, then.
Weak references and refactoring could be done to make this less of a landmine to step on later.

Reply:
Is it really a landmine? I mean, CI warns you if you forget to dispose something

Reply:
That's true, but I'd really like to avoid the extra dispose method if at all possible.